ursabot/configs.py

```diff
@@ -0,0 +1,148 @@
+        self.name = name
+        self.workers = workers
+        self.builders = builders
+        self.schedulers = schedulers
+        self.images = images or []
+        self.pollers = pollers or []
+        self.reporters = reporters or []
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__}: {self.name}>'
+
+
+def MasterConfig(title, url, webui_port, worker_port, projects,
+                 database_url=None, auth=None, authz=None, change_hook=None,
+                 secret_providers=None):
+    """Returns with the dictionary that the buildmaster pays attention to."""
+
+    def component(key):
+        return list(toolz.concat(getattr(p, key) for p in projects))
+
+    if change_hook is None:
+        hook_dialect_config = {}
+    else:
+        hook_dialect_config = change_hook._as_hook_dialect_config()
+
+    buildmaster_config = {
+        'buildbotNetUsageData': None,
+        'title': title,
+        'titleURL': url,
+        'buildbotURL': url,
+        'workers': component('workers'),
+        'builders': component('builders'),
+        'schedulers': component('schedulers'),
+        'services': component('reporters'),
+        'change_source': component('pollers'),
+        'secretsProviders': secret_providers or [],
+        'protocols': {'pb': {'port': worker_port}},
+        'db': {'db_url': database_url},
+        'www': {
+            'port': webui_port,
+            'change_hook_dialects': hook_dialect_config,
+            'plugins': {
+                'waterfall_view': {},
+                'console_view': {},
+                'grid_view': {}
+            }
+        }
+    }
+
+    # buildbot raises errors for None or empty dict values so only set of they
+    # are passed
+    if auth is not None:
+        buildmaster_config['www']['auth'] = auth
+    if authz is not None:
+        buildmaster_config['www']['authz'] = authz
+
+    return buildmaster_config
+
+
+def load_variable(config, variable):
+    """Load variable from python file
+
+    License note:
+        It is a reimplementation based on the parent GitHubStatusPush
+        from the original buildbot implementation.
+    """
+
+    config = Path(config).absolute()
+    basedir = config.parent
+
+    if not config.exists():
+        raise ConfigErrors([
+            f"configuration file '{config}' does not exist"
+        ])
+
+    try:
+        with config.open('r'):
+            pass
+    except IOError as e:
+        raise ConfigErrors([
+            f'unable to open configuration file {config}: {e}'
+        ])
+
+    log.info(f'Loading configuration from {config}')
+
+    # execute the config file
+    local_dict = {
+        'basedir': basedir.expanduser(),
+        '__file__': config
+    }
+
+    # TODO(kszucs): ask @pitrou about how to fool python to use basedir
+    #               as the base of the relative paths used within the loaded
+    #               python file
+    old_sys_path = sys.path[:]
+    sys.path.append(str(basedir))
```

Comment:
@pitrou You won't be proud at me :)
Are you aware of some black magic that `pathlib.Path('.')` (within the `__file__`) resolve to the `dirname(__file__)` instead of the current working directory?

Reply:
What do you mean with "use basedir as the base of the relative paths"? Which relative paths are you talking about?

Reply:
@pitrou Now I need to pass absolute paths [here](https://github.com/ursa-labs/ursabot/pull/140/files#diff-9002e3fc76b2e67ce086546010019f81R43-R51), previously I could use relative paths:
```py
Config.load('default.yaml', f'{env}.yaml')
```

Reply:
Perhaps you changed (or something changed) the current directory?

Reply:
So what I'm curious of, that how does `pathlib.Path().absolute()` determine the current path?

Reply:
Would it be possible to [inject another local](https://github.com/ursa-labs/ursabot/pull/140/files#diff-b9e32df4e0d6b3c92abafaf73036723fR116) to `pathlib.Path().absolute()` resolve to the [included filename](https://github.com/ursa-labs/ursabot/pull/140/files#diff-52f1c0237af74d262c495de32c7b1b05R61)?

Reply:
[os.getcwd](https://docs.python.org/3/library/os.html#os.getcwd)

Reply:
Please explain what you're trying to do.

Reply:
Not relevant anymore, but out of curiosity:

```py
# ./subdir/submod.py

from pathlib import Path

print(Path(__file__).parent.absolute())
print(Path().absolute())

```

```py
# ./main.py

from pathlib import Path

submod = Path('subdir/submod.py')
globals_ = {'__file__': str(submod.absolute())}
exec(submod.read_text(), globals_)

```

I want that the two output of `python main.py` to be the same (absolute path of `subdir/submod.py`).